File: 169.cpp
#include<iostream>
#include<unordered_map>
#include<vector>

using namespace std;

class Solution {
public:
    int majorityElement(vector<int>& nums) {
        int count = 0;
        int majority;
        for(auto n: nums){
            if(count == 0){
                majority = n;
                count = 1;
            }
            else if (majority == n){
                count++;
            }
            else{
                count--;
            }
        }
        
        return majority;
    }
};

class Solution2{
public:
	int majorityElement(vector<int>& nums){
		int limit=nums.size()/2;
		unordered_map<int,int> m;
		for(auto n:nums){
			if(++m[n]>limit)
				return n;
		}
        return -1;
	}
};
	
int main(){
	vector<int> nums={1,2,3,3,3,2,3};
	Solution s;
	cout<<s.majorityElement(nums)<<endl;

	return 0;
}

File: Python/169.py
class Solution(object):
    def majorityElement(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        d = {}
        n = len(nums)
        for num in nums:
            d[num] = d.get(num, 0) +1
            if d[num] > n/2:
                return num
        return -1



class Solution(object):
    def majorityElement(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        candidate = nums[0]
        cnt = 0
        for n in nums:
            if cnt == 0:
                candidate = n
            if candidate == n:
                cnt +=1
            else:
                cnt -=1
        return candidate
